### PR TITLE
Fix #6738: Do not replace unicode characters by LaTeX macros on unicode supported LaTeX engine

### LINKED ISSUE #6738: LaTeX output: sub/superscripts and other Unicode glyphs get "TeX-escaped" even if engine is xelatex
Arguably, [TeX-escaping](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/util/texescape.py) when using Unicode aware xelatex engine should be essentially limited to the few characters with special TeX meaning such as `$`, `{`, `\` and a few others.

For the initial report where the too extensive TeX escaping causes problems see #6733. 

However some Unicode code-points such as `ℯ` seem to have limited font support (particularly monospace, it is lacking both from DejaVu Sans Mono and GNU FreeMono), and should perhaps be retained at least for the near future in the escape list. Indeed such missing glyphs do not raise errors but only LaTeX warnings and they may escape attention.

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Feature?
- Bugfix?

### Purpose
- refs: #6738

### CODE REVIEW

File: sphinx/util/texescape.py [hunk 1]
```diff
@@ -46,6 +46,14 @@
     ('|', r'\textbar{}'),
     ('ℯ', r'e'),
     ('ⅈ', r'i'),
+    # Greek alphabet not escaped: pdflatex handles it via textalpha and inputenc
+    # OHM SIGN U+2126 is handled by LaTeX textcomp package
+]
+
+# A map Unicode characters to LaTeX representation
+# (for LaTeX engines which don't support unicode)
+unicode_tex_replacements = [
```
Comment (tk0miya) [inline sphinx/util/texescape.py:55]: I don't know what characters should be kept as-is.

Comment (jfbu) [reply sphinx/util/texescape.py:55]: I tested  the characters (one on each line below):

```
    ¶
    §
    €
    ∞
    ±
    →
    ‣
    ✓
    ✔
    ﻿
    ⎽
    –
    |
    ℯ
    ⅈ
```

with xelatex and GNU FreeFont. 

- Serif contains them all
- Sans and Mono are lacking some.

Here are the relevant warnings: (you need to see that with a font which has the glyphs...)

```
Missing character: There is no ✓ in font [FreeSans.otf]/OT:script=latn;language
=DFLT;mapping=tex-text;!
Missing character: There is no ✔ in font [FreeSans.otf]/OT:script=latn;language
=DFLT;mapping=tex-text;!
Missing character: There is no ﻿ in font [FreeSans.otf]/OT:script=latn;language
=DFLT;mapping=tex-text;!
Missing character: There is no ⎽ in font [FreeSans.otf]/OT:script=latn;language
=DFLT;mapping=tex-text;!
Missing character: There is no ⅈ in font [FreeSans.otf]/OT:script=latn;language
=DFLT;mapping=tex-text;!
 
Missing character: There is no ✓ in font [FreeMono.otf]/OT:script=latn;language
=DFLT;!
Missing character: There is no ✔ in font [FreeMono.otf]/OT:script=latn;language
=DFLT;!
Missing character: There is no ℯ in font [FreeMono.otf]/OT:script=latn;language
=DFLT;!
Missing character: There is no ⅈ in font [FreeMono.otf]/OT:script=latn;language
=DFLT;!
```

Sans is used in titles and Mono in code-blocks.

I thus think a daring but still cautious approach currently would be maintain the TeX escaping even with Unicode engines for the following characters: ``✓ , ✔ , ⎽ ,  ℯ , ⅈ`` and the ZWNBSP  (U+FEFF, also used as BYTE ORDER MARK).

But for ``¶,  §,  €,  ∞,  ±,  →,  ‣,  –,  |`` I don't see obvious reason why they should be TeX-escaped with Unicode engines.

Note 1: contrarily to web browsers, LaTeX has no easily usable concept of "fall-back" font when a glyph is missing from a font. It is up to user to add some `\DeclareUnicodeCharacter` or `\newunicodechar` to tell LaTeX that for this or that glyph, another font must be used. And making this fontfamily known to fontspec is also under user responsabillity.

Note 2: one may hesitate about escaping `ℯ` as it is availble in FreeSerif, but it is lacking from FreeMono and also from DejaVu Sans Mono. (as seen in #6733 original report). Probably same for `ⅈ`.

I am tempted to not escape them, but hope that user community will share advice on how to use `\newunicodechar` (I have not looked myself in detail and in particular how to use a fallback font for them only when needed, i.e. not in normal text which uses FreeSerif).

**Review by jfbu — APPROVED** [review]
LGTM with nits!

**Comment by jfbu** [comment]
Merging as specific Unicode characters not needing replacements can be handled later. Thanks!

**Comment by jfbu** [comment]
@tk0miya I have a problem with trying to move in particular the ∞ to `unicode_tex_replacements`. If I do that the ``test_markup.py::test_inline`` fails because ∞ is not TeX-escaped contrarily to expectation:

```
=================================== FAILURES ===================================
_ test_inline[verify-::\n\n @\u0393\\\u221e${}-None-\\begin{sphinxVerbatim}[commandchars=\\\\\\{\\}]\n@\u0393\\PYGZbs{}\\(\\infty\\)\\PYGZdl{}\\PYGZob{}\\PYGZcb{}\n\\end{sphinxVerbatim}] _
```

Other question, should

https://github.com/sphinx-doc/sphinx/blob/c2d64a9b0a7ef1fbd56c620d292e067ded462200/sphinx/ext/todo.py#L300-L308

use rather the `texescape.get_escape_func()` ?

**Comment by tk0miya** [comment]
Thank you for merging. I'd missed your comment.

About `∞` is a bug. It seems highlighting does not refer new unicode table. I'll take a look.
And about todo extension, it is also a bug.

**Pull request merged**